src/sparse/impl/KokkosSparse_bspgemm_impl.hpp

```diff
@@ -0,0 +1,207 @@
+/*
+//@HEADER
+// ************************************************************************
+//
+//                        Kokkos v. 3.0
+//       Copyright (2020) National Technology & Engineering
+//               Solutions of Sandia, LLC (NTESS).
+//
+// Under the terms of Contract DE-NA0003525 with NTESS,
+// the U.S. Government retains certain rights in this software.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are
+// met:
+//
+// 1. Redistributions of source code must retain the above copyright
+// notice, this list of conditions and the following disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright
+// notice, this list of conditions and the following disclaimer in the
+// documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the Corporation nor the names of the
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY NTESS "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL NTESS OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Siva Rajamanickam (srajama@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+*/
+
+#ifndef _KOKKOSBSPGEMMIMPL_HPP
+#define _KOKKOSBSPGEMMIMPL_HPP
+
+//#define KOKKOSKERNELS_ANALYZE_COMPRESSION
+//#define KOKKOSKERNELS_ANALYZE_MEMORYACCESS
+//#define HASHTRACK
+
+//#define TRACK_INSERTS
+//#define GPU_EXPERIMENTAL
+//#define NUMERIC_USE_STATICMEM
+//#define twostep
+
+#include "KokkosSparse_spgemm_impl.hpp"
+
+namespace KokkosSparse {
+
+namespace Impl {
+
+template <typename HandleType, typename a_row_view_t_,
+          typename a_lno_nnz_view_t_, typename a_scalar_nnz_view_t_,
+          typename b_lno_row_view_t_, typename b_lno_nnz_view_t_,
+          typename b_scalar_nnz_view_t_>
+class KokkosBSPGEMM
+    : public KokkosSPGEMM<HandleType, a_row_view_t_, a_lno_nnz_view_t_,
```

Comment:
My goal was to reuse `KokkosSPGEMM` code without modifying the original file.  
There are no virtual methods¹, but that design may not be the best one regardless, so we could switch to:
1. skip base class and have `KokkosBSPGEMM` define all required types, fields and `compute_num_pool_chunks()` method (duplicate or make it a shared utility);
2. skip extra class and implement block methods together with non-block methods inside `KokkosSPGEMM` class (which anyway serves as an umbrella for spgemm algorithms: jacobi, triangle, outer, etc.);

Would you prefer one of those or have a different suggestion ?

> ¹ In SpGEMM, only the handle has virtual table ([here](https://github.com/kokkos/kokkos-kernels/blob/403a4177774fda29426e289367de705bf9cbe0a8/src/sparse/KokkosSparse_spgemm_handle.hpp#L494-L498)), which is a bit weird since nothing derives from it.